### PR TITLE
Fix occasional OOM inside LogitsProcessor

### PR DESCRIPTION
**Pull request by pjin-nvidia**
## Motivation

Occasionally an OOM can occur inside `LogitsProcessor` where the logits are casted to fp32. This tends to occur after a few steps of RL, depending on the initial `mem_fraction_static`, due to an increase in the overall GPU memory usage during RL training.

## Modifications

This PR simply defers the fp32 cast to after the logits are sliced.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/layers/logits_processor.py [hunk 1]
```diff
@@ -514,7 +514,7 @@ def _get_logits(
             logits_buffer.copy_(logits[:, : self.config.vocab_size])
             logits = logits_buffer
         else:
-            logits = logits[:, : self.config.vocab_size].float()
+            logits = logits[:, : self.config.vocab_size]
```
Comment (ch-wan) [inline python/sglang/srt/layers/logits_processor.py:517]: we should defer the above branch too to keep code logic consistent

Comment (pjin-nvidia) [reply python/sglang/srt/layers/logits_processor.py:517]: in the branch above, the logits point to `logits_metadata.next_token_logits_buffer` which should already be pre-allocated in float32, so there is no fp32 cast to defer

see this assert: https://github.com/sgl-project/sglang/pull/9540/files/56c19973d3a50f4cb2319a982e996e5b7438bccd#diff-e7734e99e6b8d0c2279a1599685fef4b9fe4907e1ba5e0d9f240072670ae5879R513

Comment (pjin-nvidia) [reply python/sglang/srt/layers/logits_processor.py:517]: also I added a fallback cast back to fp32 before the final softcap in `_get_logits`

Comment (ch-wan) [reply python/sglang/srt/layers/logits_processor.py:517]: You are correct. But the code logic would be pretty messy. We'd better to keep output dtype consistent.

**Comment by hnyls2002** [comment]
@pjin-nvidia I agree with Chen's opinion. Could you please keep the output dtype consistent?
